src/components/tl/mlx5/mcast/tl_mlx5_mcast.h

```diff
@@ -384,6 +458,55 @@ static inline struct pp_packet* ucc_tl_mlx5_mcast_buf_get_free(ucc_tl_mlx5_mcast
     return pp;
 }
 
+static inline ucc_status_t ucc_tl_mlx5_mcast_post_user_recv_buffers(ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                                                    ucc_tl_mlx5_mcast_coll_req_t  *req,
+                                                                    int group_id, ucc_rank_t root,
+                                                                    int coll_type,
+                                                                    int count,
+                                                                    size_t offset)
+{
+    struct ibv_recv_wr *bad_wr = NULL;
+    struct ibv_recv_wr *rwr    = comm->call_rwr;
+    struct ibv_sge     *sge    = comm->call_rsgs;
+    struct pp_packet   *pp     = NULL;
+    uint32_t            i;
+
+    for (i = 0; i < count; i++) {
+        if (NULL == (pp = ucc_tl_mlx5_mcast_buf_get_free(comm))) {
+            tl_error(comm->lib, "not enought free pp packets to cover the entire message");
```

Comment:
How likely is this to happen? can't this be detected ahead of time, at collective init?

Reply:
this has never happened as we have appropriate check during collective init. But this check can be helpful if there is any memory corruption or if in the future, we change the design and have support for multiple on the flight mcast collectveis. Right now collectives are getting progress sequentially.

Reply:
So the situation is a bit tricky here.  ```ucc_tl_mlx5_mcast_buf_get_free``` is getting called 3 other places. Over there, if there is not enough pp packets, we can just return and try again later once we have enough packets available. So it is not an issue nor an error. 

However, here in this piece of code, if we run out of pp packets, then it usually means something went wrong with QP as we must have enough pp packet when we call this function here (we make sure of that before getting into here).
 Returning error instead of assert can better demonstrate what went wrong and we can add some error handling on the upper layers (like possibly draining the recv queue and trying again?)